### ISSUE TITLE
Share calendar to users or groups with "show only as busy".

### ISSUE THREAD

**Issue by phlegx** [nextcloud/server#11214]
### Steps to reproduce
1. Click on calendar app
2. Click on share calendar
3. Select user or group

### Expected behaviour
After the steps to reproduce the behavior I want to choose an option "show only as busy". So, that the selected user or group can see my complete calendar but without details. This option is missing.

I know that it is possible for a single event to "show only as busy". But what if I want to have this for a **specific user or group** for the **complete** calendar?

### Actual behaviour
After the steps to reproduce the behavior there is only an option "can edit" for the chosen user or group.

### Server configuration

**Nextcloud version:** 13.0.6
**Nextcloud calendar version:** 1.6.1

**Comment by MorrisJobke** [comment]
cc @georgehrke

**Comment by stefanoco** [comment]
any further activity on this one? We badly need this feature (and was surprised not to see other similar comments here). Open to give bounty for this.

**Comment by georgehrke** [comment]
Nope, biggest help here would be someone sending a pull-request. Happy to give instructions.

**Comment by stefanoco** [comment]
Ok I'll try to understand if it's possible for me to help (in a few days) and submit a pull request, thanks.

**Comment by jfrederickson** [comment]
Ideally it'd be possible to do this for calendar links as well. (And to be able to generate links for both full access and free/busy simultaneously.) Google Calendar for example only lets you do one or the other at any given time.

**Comment by nolens** [comment]
I badly need this feature and I have just set a bounty on this.

**Comment by nolens** [comment]
@stefanoco 
Would you like add to my bounty on this?  I have a couple of freelance programers who I am working with on other projects, they might be able to contribute on short notice.... 

@georgehrke 
Could we setup a group chat somewhere for us and the programers, to discuss scope and for you to give clear directions? Could make life easier, in the end...

**Comment by georgehrke** [comment]
> Could we setup a group chat somewhere for us and the programers, to discuss scope and for you to give clear directions? Could make life easier, in the end...

Sure, here you go: https://cloud.nextcloud.com/call/pjvxs42t

**Comment by brylie** [comment]
This would be really important to avoid accidentally publishing event details to a shared calendar. Right now, events seem to default to "when shared, show full event". This may accidentally expose private information when creating an event from a phone or app that doesn't expose the "when shared show" field.

Is it possible to change the default value for "when shared show" field to "when shared, show only busy"?

**Comment by biva** [comment]
Hello, any news on this feature? @stefanoco would you still be ready to work on this great feature? Thank you in advance!

**Comment by stefanoco** [comment]
Unfort I didn't have time to look after this and don't see any possibilities in the near future so I'd like to add to the bounty. Anyone willing to solve?

**Comment by biva** [comment]
As a workaround, and within our organisation, I use the free-busy feature of caldav within Thunderbird: https://docs.nextcloud.com/server/latest/admin_manual/groupware/calendar.html#freebusy

**Comment by aresius** [comment]
> As a workaround, and within our organisation, I use the free-busy feature of caldav within Thunderbird: https://docs.nextcloud.com/server/latest/admin_manual/groupware/calendar.html#freebusy

This sounds interesting, but I do not understand directly how that works. Could you please briefly give an idea of your setup?

**Comment by szaimen** [comment]
I suppose this is still valid on NC21.0.2?

**Comment by trevor87** [comment]
> As a workaround, and within our organisation, I use the free-busy feature of caldav within Thunderbird: https://docs.nextcloud.com/server/latest/admin_manual/groupware/calendar.html#freebusy

Did anyone figure out, how this works? Unfortunately is very little information online about this workaround… All I could find out so far is that this feature is probably used when creating a new appointment and then pressing the "Invite Attendees" button.

**Comment by biva** [comment]
This is it @trevor87 ! Nothing more to do

**Comment by trevor87** [comment]
OK, thanks I think I figured it out now. So for everyone who runs into this problem, the workaround is simply the following:

If you use thunderbird and want to see whether your colleagues are available just doubleclick in the Thunderbird Calendar to create a new event and then in the next view click on "Invite Attendees". When you enter a mail address of someone who is also using your nextcloud instance, it will show when the other person has appointments. In my case for some reason it shows the other people's appointments as "no information", which I find rather confusing.

Otherwise: Still trying to figure the rest out. Hope this helps!

**Comment by nursoda** [comment]
@trevor87 Thanks. That is about what Outlook does. My issue with this approach is that I don't only need my employees using NC and TB to be able to see when they may invite me, but also my customers. Therefore I'd like two things to be available: a NC web interface "free only" view for those who have an account on my server, and a publicly sharable calendar (similar to what Google Calendar provides). But maybe this is not the right place to ask for the latter.

**Comment by qq7te** [comment]
I also would like to stress what @nursoda mentioned recently: I also need it to be able to publish a calendar externally, but I want all the time slots to simply say "busy" so that I can share my calendar anywhere with clients without revealing what I'm doing

**Comment by ChristophWurst** [comment]
@qq7te we'll soon look into https://github.com/nextcloud/calendar/issues/1023. This might be necessary for that but maybe it's not. I'll let you know if we have to work on this specific ticket.

**Comment by nursoda** [comment]
> We'll soon look into [nextcloud/calendar#1023](https://github.com/nextcloud/calendar/issues/1023). This might be necessary for that but maybe it's not.

Well, "appointments" were implemented, but nothing to help "share/show busy only".

I try to sum up my understanding of this issue and my needs in more detail: We need a new? endpoint that returns only pruned events matching restrictions which a user should be able to set. Such filters might be

- only show events that are at least partly within the time frame set in `/settings/user/groupware`
- don't show certain event types such as day, private, cancelled or tentative events (or only show "confirmed" events?)
- don't show events marked as "free" (or only show events marked as "busy"?)

The OP's request was to be able to share such restricted view _within_ the nextcloud instance. I second that. This probably would be a checkbox "show only as busy" in the calendar's prefs. Since such restricted calendar view should be available to other clients such as Thunderbird, too, just a restricted view within the calendar app doesn't seem sufficient from a security perspective.

My usecase is an enhancement to the OP's request: In the Calendar app, a user should be able to "Share free/busy link" (below "Share link"); an URL that renders a web view with buttons to switch view and move in time, but without e.g. the possibility to download that calendar, that renders the filtered/pruned events.

For me, any web view would suffice. A further enhancement would be have a user setting that either renders a full-fledged view with top bar (cloud and/or calendar name, etc.) or just a minimal (embed-style) view.

**Comment by paradeiser** [comment]
I was happy to see this OP request, as this is exactly what's missing to make Nexcloud the Google Calendar successor.

For now, it is not possible to have a private Calendar shared with team-mates as "show only busy times".
(A second "private" calendar is needed as CalendarApps using CalDAV do not support the per-appointment-setting "show full details / as busy / not at all").

Is there any active development for this feature?
(Or a workaround to for private entries via CalendarApps?)

Thanks everyone for contributing to Nextcloud!
paradeiser

**Comment by paradeiser** [comment]
I just wonder if this would be better off at https://github.com/nextcloud/calendar ?

**Comment by krombel** [comment]
There is https://github.com/nextcloud/calendar/issues/1044 and https://github.com/nextcloud/calendar/issues/725 which point to this issue

**Comment by Jaxom99** [comment]
Maybe a workaround could be found with the Calcleaner project : see my open issue here https://github.com/flozz/calcleaner/issues/3

In short: automate a caldav tool (in python) to add "visibility settings" to created events (CLASS:CONFIDENTIAL). Thus a standard NC share of the calendar would show only date/time.

**Comment by sm8ps** [comment]
Judging from the many requests spread widely over different threads this seems to be a much sought-after feature as well as one that is not very likely to be added to Nextcloud very soon. 

Since I need the functionality myself, I have cobbled together a [script](https://github.com/sm8ps/shadowCalDAV) that makes a shadow copy of an ("internal") calendar to a "external" (shadow) calendar that can be published e.g. to the web.

This is no ideal solution by any means but it has provided useful so far. Feedback and suggestions welcome!

**Comment by section1** [comment]
I have similar issue ...we want to share calendar inside NC instance and we want that always the shared calendar show the events as 'Busy'.. i know that exist de free/busy option when you create and event...but sometimes is more easy that multiple users share the calendar to me and i can found a slot very fast in the 'calendar page'.
i don't need to create an event, add attendants, find free slot for everyone etc.. 

I'm testing this stupid workaround that i make:

Edit file apps/dav/lib/CalDAV/CalendarObject.php search get() function.. in this version i have something like this:

~~~
                // shows as busy if event is declared confidential
                if ($this->objectData['classification'] === CalDavBackend::CLASSIFICATION_CONFIDENTIAL) {
                        $this->createConfidentialObject($vObject);
                }

~~~

if you remove this if clause (keeping ` $this->createConfidentialObject($vObject);` obviously)  you'll see all events as Busy at shared calendars.

If someone knows if this is a very bad idea please tell me :D  , I can't find any bugs or problems so far.

**Comment by wargreen** [comment]
Hi,
The per-event and per-calendar workaround isn't really the same feature.
In my use case, i need to share the free/busy for users that use multiple calendars for multiple projects. Currently, the free/busy info are taken from all the cal a user own (if i have good information), so it is far most complete than just share one cal.
Is the per-user free/busy share the same issue, from an internal Nextcloud point-of-view ?

**Comment by wargreen** [comment]
For detail my previous comment (i've found it when searching again for this feature :) :
My use case : i work with many music / theater shows. One calendar each (that i'm or not owner), but not all want to use NC. I also have many other shared calendars for associative life etc, that don't should be used for free/busy...
So, many frends, productions and tour managers need to know when i'm available (sometime from goo**e things).

That i can imagine a good way :

- Have a "Share links" section in left calendar panel
  - That list all shared links and a button "Create share"
    - That open a popup for :
      - select which calendar to use (in all viewable calendars by user)
      - select to show full or free/busy
      - manage share (same as the share panel for files) 

The "confidential trick" said before can't work for me (already tried) because if i add an event in a shared cal, only the cal owner can see it.
In my case, a "personal / user point-of view" of the free/busy information, and all information are not necessarily in one calendar.

I think that the current state of calendar sharing work well for a monolitic organization (one NC instance / one orga) but is complicated for one-by-one (inter-users or user-to-external-world) organization.

Thanks a lot for all the work in all cases !

**Comment by section1** [comment]
Yes, I'm agreed, that would be a nice and important('must have') feature. I want to add that the workaround above have a bug with recurrent events. i make an issue here: https://github.com/nextcloud/calendar/issues/5555

**Comment by tetefolle** [comment]
+1

**Comment by bentuna** [comment]
Is this still open for NC28?

**Comment by mdrobisch** [comment]
To whom it may concern.

We have a Nextcloud setup in use and a lot of different user clients including MS Outlook and Thunderbird (with CalDav).

Finding free meeting slots was constantly hard to manage because of hidden "private" events of MS Outlook user. The behaviour is described a lot in different issues here.

With the hint of @section1 (thanks alot here) we managed to extend his solution to show all events as "Busy"-timeslot. So this includes "confidential" and "private" types.

Here is the solution:

In `/nextcloud/apps/dav/lib/CalDAV/CalendarObject.php` we handled private events separatly with 

```
class CalendarObject extends \Sabre\CalDAV\CalendarObject {

	public function get() {

		// ... other code of the function

		// shows as busy if event is declared confidential
		if ($this->objectData['classification'] === CalDavBackend::CLASSIFICATION_CONFIDENTIAL) {
				$this->createConfidentialObject($vObject);
		}

		// code added here !!!!!!!!!!!!!!!!!!!!
		// shows as busy if event is declared private
		if ($this->objectData['classification'] === CalDavBackend::CLASSIFICATION_PRIVATE) {
				$this->createConfidentialObject($vObject);
		}

		// ... other code of the function

	}

```  

To get this working we had to drop the filters for private event in `/nextcloud/apps/dav/lib/CalDAV/Calendar.php` as well.

We changed the behaviour in the following functions.

* public function getChild($name) { ... }
* public function getChildren() { ... }
* public function getMultipleChildren(array $paths) { ... }
* public function childExists($name) { ... }

Simply by commenting out the `CLASSIFICATION_PRIVATE` section for every section/function like below:

```
	//if ($obj['classification'] === CalDavBackend::CLASSIFICATION_PRIVATE && $this->isShared()) {
	//      return false;
	//}
```

So the filtering is done in the `CalendarObject.php` only

If you interested in this. You can change this in your live system directly without any restart. 
Please be careful to do not mess up with the other code. Please also tell your thunderbird users about the visibility of "private" events as well. 

As a result nextcloud will show every event in the Nextcloud calendar. It works even with the sync by Thunderbird and Outlook (CalDav) showing "Busy" slots. This does the trick for us and we are really have with the solution.

For your decission to patch, please mention that this is a workaorround only and it may leak out data like e.g. described in https://github.com/nextcloud/calendar/issues/5555 

If anyone can show me how to make this an option in the admin setting (or even calendar wise) i am interested to make a Pull-request to have a general solution for all nextcloud users.

**Comment by kaystrobach** [comment]
@mdrobisch thank you for your input, this was not the intention of the ticket.

We want to modify the sharing options in a way, so that the sharing user can decide to hide all events, no matter, what kind and just provide the busy information. Modifying each and every event can cause quite a huge security risk.

**Comment by mdrobisch** [comment]
Sure thing. Maybe https://github.com/nextcloud/calendar/issues/4782 fits better, or even one of the other issues. They imho describe all some different aspects of the same topic. So for me it is hard to figure out where my findings placed the best. Hopefully my workarround here can help people that als want a more "MS Exchange" like behaviour of showing/hiding private events.

If i can help please let me know.

**Comment by mdrobisch** [comment]
> Modifying each and every event can cause quite a huge security risk.

Please let me put one addition here to clear things out. My approach do not alter events. It is just about change the filter to let "private" events to be shown as "busy" events here.

**Comment by kaystrobach** [comment]
yes totally get that, but all non "private" events will be shown - right?

**Comment by kaystrobach** [comment]
and yes https://github.com/nextcloud/calendar/issues/4782 seems to be the very same issue ...

**Comment by mdrobisch** [comment]
> yes totally get that, but all non "private" events will be shown - right?

Yes, all events that are not "private" nor "confidential" are shown/shared as normal (with all informations).

Thank you for the hint. You are right. This solution fits best with https://github.com/nextcloud/calendar/issues/4782 not with this one here.

**Comment by miaulalala** [comment]
Related: https://github.com/nextcloud/calendar/issues/6247

**Comment by levino** [comment]
What I am missing is a "default template" for new entries where I can set the visibility. Currently it seems to be "if shared show everything". I can select "if shared show only busy" which is good. However I would like this value to be the default for a new entry so I do not forget / get "plausible deniability" :).

**Comment by IcedRooibos** [comment]
Are there any workarounds for this at the moment? I'm sure I'll eventually forget to check the box that says keep the event private and having it done for the whole calendar link would be great.

**Comment by kaystrobach** [comment]
we use bots to copy empty events to a seperate calender, but this is not what we intended.

**Comment by mdlaat** [comment]
> Are there any workarounds for this at the moment? I'm sure I'll eventually forget to check the box that says keep the event private and having it done for the whole calendar link would be great.

I created a simple 'proxy' in php, which returns the calendar but with the events anonymised (attendees removed, location, removed, appointmentname is 'busy', etc).

**Comment by levino** [comment]
Can this not be implemented in 5 minutes by copilot coding agent? Can someone assign them this issue?

**Comment by levino** [comment]
I mean it. Just try it. It really works. I am using it in all my repos and it is pretty good. Worth a try and costs like a dollar per feature.

**Comment by bwurst** [comment]
@AndyScherzinger did you possibly close the wrong bug report?

This report is about a demanded new feature for sharing a whole calendar as "only show me busy". 
I'd like to have a consumer for my calendar that should never see the details of events but others should.

**Comment by AndyScherzinger** [comment]
> did you possibly close the wrong bug report?

Got auto-closed by merging the referenced PR. @st3iny wrong referenced issue?

**Comment by st3iny** [comment]
Yeah, this issue was referenced wrongly.

**Comment by JlincFM** [comment]
Question: To support this, would something as simple as the below work?

- Add a checkbox option to the calendar settings of "When shared show only busy" (default unchecked)
  - When checked:
    - If a new event is created: automatically set the event to "When shared show only busy"
    - If an event is updated: use whatever visibility value comes through
  - When unchecked:
    - Current behavior

**Comment by kaystrobach** [comment]
No

We need different views for different purposes.

* Busy / non busy view for people / apps / devices who we don’t trust
* Normal behavior, where we can decide on event base

So basically share options are the way to go, something to select how detailed event info is shared like all, minimal (title only), just timeslot

**Comment by JlincFM** [comment]
Makes sense. I was trying to think of this in baby steps (small PRs). I.E.:


- **Phase 1**: Add the calendar preference for default when new events are added
  - **Note**: This looks straight forward, solves many of the above use cases, and is a nice to have for phase 2 anyway.
- **Phase 2**: Full sharing preferences that override the event settings.

**Comment by doppelgrau** [comment]
I wrote now a small(?) python script that "solves" this problem for me. It "copies" the entries from some calenders into another one. Either just with predefined Title/Summary or copying the original title/summary after the predefined one (e.g. if some details shouldn't be shared that might be in the notes) but it's ok to get an overall idea what the appointment is about.

```python
from caldav import error, get_davclient
import calendar
from datetime import datetime, timedelta
import icalendar


# Calendars with title
# Name (used in the copied events): URL
calendars_with_title = {\
        "Private":"https://nextcloud/dav/calendars/user/calendar_name/"
    }
# Calendars with only "busy"
# Name (used in the copied events): URL
calendars_only_busy = {\
        "Work":"https://nextcloud/dav/calendars/user/calendar_name/"
    }
# Destination calendar
target_calendar = "https://nextcloud/dav/calendars/user/calendar_name/"

#
# Helpers
#
def replace_calendar_url_by_object(url, calendars):
    for calendar in calendars:
        if url == calendar.url:
            return calendar

def replace_calendar_urls_in_dictionary_by_object(dictionary, calendars):
    for key in dictionary:
        dictionary[key] = replace_calendar_url_by_object(dictionary[key], calendars)
    return dictionary

def delete_old_entries(calendar, min_age):
    old_events = calendar.search(
        start=datetime.now() + timedelta(days= -2 * min_age),
        end=datetime.now() + timedelta(days= -1 * min_age),
        event=True,
        expand=True,
    )
    for event in old_events:
        event.delete()
    
def get_next_event_from_iterator(iterator):
    try:
        event = iterator.__next__()
        return event
    except:
        return None
    
def get_event(raw_event):
    for component in raw_event.get_icalendar_instance().subcomponents:
        if isinstance(component, icalendar.cal.Event):
            return component

def sync_calender(source, target, name, copy_summary=False):
    # get events in the relevant intervall
    source_events = source.search(
        start=datetime.now(),
        end=datetime.now() + timedelta(days=270),
        event=True,
        expand=True,
    )
    target_events = target.search(
        start=datetime.now(),
        end=datetime.now() + timedelta(days=270),
        event=True,
        expand=True,
    )
    
    # sort according to start time
    source_events.sort(key=lambda self: calendar.timegm(get_event(self)["DTSTART"].dt.timetuple()))
    target_events.sort(key=lambda self: calendar.timegm(get_event(self)["DTSTART"].dt.timetuple()))

    # get iterators
    source_iter = source_events.__iter__()
    target_iter = target_events.__iter__()
    
    # get first event
    current_source_event = get_next_event_from_iterator(source_iter)
    current_target_event = get_next_event_from_iterator(target_iter)
    
    # both calendars still have events
    while current_source_event != None and current_target_event != None:
        source_vevent = get_event(current_source_event)
        target_vevent = get_event(current_target_event)
        # make sure the event in the target calender is for the source, else proceed forward
        if target_vevent["SUMMARY"].find(name) != 0:
            current_target_event = get_next_event_from_iterator(target_iter)
            continue
        # we have two events for the source calender, now lets check if they are sufficient identical
        if copy_summary:
            target_summary = name + ": " + source_vevent["SUMMARY"]
        else:
            target_summary = name
        if target_vevent["SUMMARY"] == target_summary \
           and source_vevent["DTSTART"].dt == target_vevent["DTSTART"].dt \
           and source_vevent["DTEND"].dt == target_vevent["DTEND"].dt:
            # entries are suffienctly identical, proced to the next ones
            current_source_event = get_next_event_from_iterator(source_iter)
            current_target_event = get_next_event_from_iterator(target_iter)
            continue
        # the events are different
        # first case, the one in the target starts earlier (or at the same time) => delete that one
        if source_vevent["DTSTART"].dt >= target_vevent["DTSTART"].dt:
            current_target_event.delete()
            current_target_event = get_next_event_from_iterator(target_iter)
            continue
        # second case, the one in the source starts earlier, => create a new event
        if source_vevent["DTSTART"].dt < target_vevent["DTSTART"].dt:
            target.add_event(
                        dtstart=source_vevent["DTSTART"],
                        dtend=source_vevent["DTEND"],
                        summary=target_summary,)
            current_source_event = get_next_event_from_iterator(source_iter)
            continue
    
    # While loop finished, one or both calendars have no remaining events
    assert current_source_event == None or current_target_event == None

    # target still has some events, delete if matching the source
    while current_target_event != None:
        target_vevent = get_event(current_target_event)
        # originally from the source calendar
        if target_vevent["SUMMARY"].find(name) == 0:
            current_target_event.delete()
        # check next
        current_target_event = get_next_event_from_iterator(target_iter)

    # source has still some events, add those to the target
    while current_source_event != None:
        source_vevent = get_event(current_source_event)
        # originally from the source calendar
        if copy_summary:
            target_summary = name + ": " + source_vevent["SUMMARY"]
        else:
            target_summary = name
        target.add_event(
                    dtstart=source_vevent["DTSTART"],
                    dtend=source_vevent["DTEND"],
                    summary=target_summary,)
        # check next
        current_source_event = get_next_event_from_iterator(source_iter)

## The client object can be used as a context manager, like this:
with get_davclient() as client:
    # Connecting to the caldav server
    my_principal = client.principal()

    # The principals calendars can be fetched like this:
    calendars = my_principal.get_calendars()
    
    # replace the URLs by more conventient objects
    target_calendar = replace_calendar_url_by_object(target_calendar, calendars)
    calendars_with_title = replace_calendar_urls_in_dictionary_by_object(calendars_with_title, calendars)
    calendars_only_busy = replace_calendar_urls_in_dictionary_by_object(calendars_only_busy, calendars)

    # delete old entries
    delete_old_entries(target_calendar, 90)
    
    # sync "busy"
    for calendar_name in calendars_only_busy:
        sync_calender(calendars_only_busy[calendar_name], target_calendar, calendar_name)
    # sync "summary"
    for calendar_name in calendars_with_title:
        sync_calender(calendars_with_title[calendar_name], target_calendar, calendar_name, True)
```
Quite sure it's not perfect, but so far works "good enough". Assumes the `CALDAV_USERNAME`, `CALDAV_PASSWORD` and the `CALDAV_URL` (like `CALDAV_URL=https://nextcloud/remote.php/dav`) are set as environment variables.
